### PR TITLE
Prevent redundant token rate updates

### LINKED ISSUE #3605: [assets-controllers] Prevent redundant token rate updates
The `TokenRatesController` will unnecessarily update the token rates in some situations. Efforts have been made in the past to reduce the occurrences of this, but a few remain.

One example is that updates to `allDetectedTokens` can trigger a token rate update, even when the list of token addresses actually used remains unchanged. The code as written assumes that `allDetectedTokens` is always updated immutably, but that doesn't appear to be the case (and even if it was, changes to metadata would still unnecessarily trigger updates).

We can solve this by calculating the list of token addresses each time we notice a state change that might require an update. We can compare the list of addresses before and after the change to determine whether an update is needed. This should effectively prevent all unnecessary token rate updates from being triggered.

### PR DESCRIPTION
**Pull request by mcmire**
## Explanation

The token rates controller fetches rates for certain collections of tokens obtained via the tokens controller. If any of these collections change, then new rates need to be fetched.

Unfortunately, because the check to determine whether these collections have changed is simplistic, new rates are fetched every single time the tokens controller state is changed.

To address this, we can consider two things:

- The token rates controller only uses the addresses of tokens in these collections to fetch rates, and it also normalizes those addresses when updating state.
- These collections of tokens may contain duplicates.

So to reduce the number of times rates are fetched and to optimize each request, this commit updates the tokens state callback to assemble the aforementioned token collections into a deduplicated, normalized list of token addresses, and then it checks to see if this list has any changes to know whether or not to re-fetch rates.

## References

Fixes #3605.

### `@metamask/assets-controllers`

- **FIXED**: Fix TokenRatesController to prevent redundant token rate updates when tokens change
  - Previously, token rates would be re-fetched for the globally selected network on all TokensController state changes, but now token rates are always performed for a deduplicated and normalized set of addresses, and changes to this set determine whether rates should be re-fetched.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 1]
```diff
@@ -482,7 +490,7 @@ export class TokenRatesController extends PollingControllerV1<
       (obj, [tokenContractAddress, tokenPrice]) => {
         return {
           ...obj,
-          [toChecksumHexAddress(tokenContractAddress)]: tokenPrice.value,
+          [tokenContractAddress]: tokenPrice.value,
```
Comment (mcmire) [inline packages/assets-controllers/src/TokenRatesController.ts:493]: By the time we get here, `tokenContractAddress` is already normalized (thanks to `#getTokenAddresses`).

**Review by cryptodev-2s — APPROVED** [review]
LGTM!

File: packages/assets-controllers/src/TokenRatesController.ts [hunk 2]
```diff
@@ -276,7 +278,13 @@ export class TokenRatesController extends PollingControllerV1<
     const detectedTokens =
       allDetectedTokens[chainId]?.[this.config.selectedAddress] || [];
 
-    return [...tokens, ...detectedTokens].map((token) => toHex(token.address));
+    return [
+      ...new Set(
+        [...tokens, ...detectedTokens].map((token) =>
+          toHex(toChecksumHexAddress(token.address)),
```
Comment (Gudahtt) [inline packages/assets-controllers/src/TokenRatesController.ts:284]: is the `toHex` call redundant alongside `toChecksumHexAddress`?

Comment (mcmire) [reply packages/assets-controllers/src/TokenRatesController.ts:284]: It's not — unfortunately `toChecksumHexAddress` (located in `controller-utils`) doesn't use a type predicate to assert that its return value is a `Hex`. `toHex` (which is from `@metamask/utils`) does, though. So it's redundant in the sense that it doesn't do anything, but it does convert this to a `Hex`.

**Pull request merged**